crucible/src/Lang/Crucible/Backend.hs

```diff
@@ -244,26 +249,43 @@ class IsBoolSolver sym where
   resetAssumptionState :: sym -> IO ()
   resetAssumptionState sym = restoreAssumptionState sym PG.emptyGoalCollector
 
--- | Add a proof obligation for the given predicate, and then assume it.
+assertThenAssumeConfigOption :: ConfigOption BaseBoolType
+assertThenAssumeConfigOption = configOption knownRepr "assertThenAssume"
+
+assertThenAssumeOption :: ConfigDesc
+assertThenAssumeOption = mkOpt
+  assertThenAssumeConfigOption
+  boolOptSty
+  (Just (PP.text "Assume a predicate after asserting it."))
+  (Just (ConcreteBool False))
+
+-- | Add a proof obligation for the given predicate, and then assume it
+-- (when the assertThenAssume option is true).
 -- Note that assuming the prediate might cause the current execution
 -- path to abort, if we happened to assume something that is obviously false.
 addAssertion ::
   (IsExprBuilder sym, IsBoolSolver sym) =>
   sym -> Assertion sym -> IO ()
 addAssertion sym a@(AS.LabeledPred p msg) =
   do addProofObligation sym a
-     addAssumption sym (AS.LabeledPred p (AssumingNoError msg))
+     assert_then_assume_opt <- getOpt
+       =<< getOptionSetting assertThenAssumeConfigOption (getConfiguration sym)
+     when assert_then_assume_opt $
+       addAssumption sym (AS.LabeledPred p (AssumingNoError msg))
```

Comment:
We could probably benefit from sharing this code between `addAssertion` and `addDurableAssertion`.

Reply:
👍